marlowe-spec-test/src/Marlowe/Spec/ClientProcess.hs

```diff
@@ -14,46 +15,69 @@ where
-import System.IO (hSetBuffering, hGetLine,BufferMode(..))
+import System.IO (hSetBuffering, hGetLine,BufferMode(..), openFile)
 import qualified Data.Aeson.Types as JSON
 import qualified Data.Aeson as JSON
 import GHC.IO.Handle (Handle)
 import Test.Tasty (TestTree, withResource, askOption)
 import Marlowe.Spec.Interpret (Request(..), Response(..))
 import Data.Data (Typeable)
-import Test.Tasty.Options (IsOption(..), mkOptionCLParser, mkFlagCLParser, safeReadBool)
+import Test.Tasty.Options (IsOption(..), mkOptionCLParser, safeRead)
 import Options.Applicative (short)
 import Options.Applicative.Builder (metavar)
-import Test.Tasty.HUnit (assertFailure)
-import GHC.Base (when)
+import Control.Concurrent.STM.TQueue (TQueue, newTQueue, readTQueue, writeTQueue )
+import Control.Monad (forM_, forM)
+import Control.Concurrent.STM (atomically)
+import System.Timeout (timeout)
+import GHC.IO.IOMode (IOMode(..))
+import Data.Functor (void)
+
+type ClientProcessId = Int
 
 data ClientProcess =
   ClientProcess
-    { stdInH  :: Handle
-    , stdOutH :: Handle
-    , procH   :: ProcessHandle
-    , traceCP :: Bool
+    { stdInH  :: Handle           -- ^ File handle to send requests to a Client Process
+    , stdOutH :: Handle           -- ^ File handle to receive responses from the Client Process
+    , traceH  :: Maybe Handle     -- ^ File handle to put debug information
+    , procH   :: ProcessHandle    -- ^ Process handle to control the client process
+    , cpId    :: ClientProcessId  -- ^ Internal client process id
+    }
+
+data ClientProcessPool =
+  ClientProcessPool
+    { threadPool      :: TQueue ClientProcess
+    , cliCmd          :: FilePath
     }
 
 sendJsonRequest :: ClientProcess -> Request JSON.Value -> IO ()
 sendJsonRequest cp r = do
-  let
-    h = stdInH cp
-    encodedReq = JSON.encode r
-  hPutStrLn h "```"
-  hPutStrLn h encodedReq
-  hPutStrLn h "```"
-  when (traceCP cp) $ putStrLn $ "Send JSON request: " ++ C.unpack encodedReq
-
-parseJsonResponse :: ClientProcess -> IO (Maybe (Response JSON.Value))
+    let
+      encodedReq = JSON.encode r
+      h = stdInH cp
+      debugH = traceH cp
+      pid = cpId cp
+
+    hPutStrLn h "```"
+    hPutStrLn h encodedReq
+    hPutStrLn h "```"
+    debug debugH $ show pid ++ " - Send JSON request: " ++ C.unpack encodedReq
+
+parseJsonResponse :: ClientProcess -> IO (Response JSON.Value)
 parseJsonResponse cp = do
-  let
-    h = stdOutH cp
-  line <- hGetLine h
-  let
-    mRes = JSON.decode $ C.pack line
-  when (traceCP cp) $ putStrLn $ "Parse JSON response: " ++ show mRes
-  return mRes
+    let
+      h = stdOutH cp
+      debugH = traceH cp
+      pid = cpId cp
+    mLine <- timeout (5000000) $ hGetLine h
+    let
+      res = case mLine of
+        Nothing -> RequestTimeOut
+        Just line -> case JSON.decode $ C.pack line of
+          Nothing -> ResponseFailure  "Could not parse response from the client cli"
+          Just r -> r
```

Comment:
```suggestion
      tryParseRes = fromMaybe (ResponseFailure "Could not parse response from the client cli")
                      . JSON.decode . C.pack
      res = maybe RequestTimeout tryParseRes mLine
```
I thought it would be clearer, but not sure anymore, your call 🤷